nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/dispatcher/DatagramChannelDispatcher.java

```diff
@@ -0,0 +1,164 @@
+            }
+        }
+        datagramChannel.socket().bind(new InetSocketAddress(port));
+        selector = Selector.open();
+        datagramChannel.register(selector, SelectionKey.OP_READ);
+    }
+
+    @Override
+    public void run() {
+        final ByteBuffer buffer = bufferPool.poll();
+        while (!stopped) {
+            try {
+                int selected = selector.select();
+                if (selected > 0){
+                    Iterator<SelectionKey> selectorKeys = selector.selectedKeys().iterator();
+                    while (selectorKeys.hasNext()) {
+                        SelectionKey key = selectorKeys.next();
+                        selectorKeys.remove();
+                        if (!key.isValid()) {
+                            continue;
+                        }
+                        DatagramChannel channel = (DatagramChannel) key.channel();
+                        SocketAddress socketAddress;
+                        buffer.clear();
+                        while (!stopped && (socketAddress = channel.receive(buffer)) != null) {
+                            String sender = "";
+                            if (socketAddress instanceof InetSocketAddress) {
+                                sender = ((InetSocketAddress) socketAddress).getAddress().toString();
+                            }
+
+                            // create a byte array from the buffer
+                            buffer.flip();
+                            byte bytes[] = new byte[buffer.limit()];
+                            buffer.get(bytes, 0, buffer.limit());
+
+                            final Map<String,String> metadata = EventFactoryUtil.createMapWithSender(sender);
+                            final E event = eventFactory.create(bytes, metadata, null);
+
+                            // queue the raw message with the sender, block until space is available
+                            events.put(event);
+                            buffer.clear();
+                        }
+                    }
+                }
+            } catch (InterruptedException e) {
+                stopped = true;
+            } catch (IOException e) {
+                logger.error("Error reading from DatagramChannel", e);
+            }
+        }
+
+        if (buffer != null) {
+            try {
+                bufferPool.put(buffer);
+            } catch (InterruptedException e) {
+                // nothing to do here
+            }
+        }
+    }
+
+    @Override
+    public int getPort() {
+        return datagramChannel == null ? 0 : datagramChannel.socket().getLocalPort();
+    }
+
+    @Override
+    public void stop() {
+        selector.wakeup();
+        stopped = true;
+    }
+
+    @Override
+    public void close() {
+        IOUtils.closeQuietly(selector);
+        IOUtils.closeQuietly(datagramChannel);
+    }
+
+    @Override
+    public void completeConnection(SelectionKey key) {
```

Comment:
since these two are not useful for this, should the interface ChannelDispatcher include these? Or should there be an AsyncChannelDispatcher interface that extends ChannelDispatcher?